ddl/ddl_api.go

```diff
@@ -5118,3 +5122,104 @@ func (d *ddl) AlterIndexVisibility(ctx sessionctx.Context, ident ast.Ident, inde
 	err = d.callHookOnChanged(err)
 	return errors.Trace(err)
 }
+
+func validPlacementSpecs(specs []*ast.PlacementSpec) ([]*placement.Rule, error) {
+	rules := make([]*placement.Rule, 0, len(specs))
+	for k, spec := range specs {
+		rule := &placement.Rule{
+			ID:       "default_rule",
+			Count:    int(spec.Replicas),
+			Override: false,
+		}
+
+		switch spec.Tp {
+		case ast.PlacementAdd:
+		default:
+			return rules, errors.Errorf("invalid placement spec[%d], unknown action type: %d", k, spec.Tp)
+		}
+
+		switch spec.Role {
+		case ast.PlacementRoleFollower:
+			rule.Role = placement.Follower
+		case ast.PlacementRoleLeader:
+			rule.Role = placement.Leader
+		case ast.PlacementRoleLearner:
+			rule.Role = placement.Learner
+		case ast.PlacementRoleVoter:
+			rule.Role = placement.Voter
+		default:
+			return rules, errors.Errorf("invalid placement spec[%d], unknown role: %d", k, spec.Role)
+		}
+
+		for _, label := range strings.Split(spec.Constraints, ",") {
+			var op placement.LabelConstraintOp
+			switch label[0] {
+			case '+':
+				op = placement.In
+			case '-':
+				op = placement.NotIn
+			default:
+				return rules, errors.Errorf("invalid placement spec[%d], unknown operation: %c", k, label[0])
+			}
+
+			idx := strings.IndexByte(label, '=')
+			if idx+1 >= len(label) {
+				return rules, errors.Errorf("invalid placement spec[%d], empty label value: %s", k, label)
+			}
+
+			rule.LabelConstraints = append(rule.LabelConstraints, placement.LabelConstraint{
+				Key:    label[1:idx],
+				Op:     op,
+				Values: []string{label[idx+1:]},
```

Comment:
`label[idx+1:]` cannot be empty.

Reply:
https://github.com/pingcap/tidb/pull/18611/commits/93eccc87eed6ea794ee6a0d5b8e1e63ecc49fa36 should guarantee the safety.